### DISCUSSION TITLE
Documentation and example as part of each PR

### DISCUSSION THREAD

**Discussion by maeste** [wise-agents/wise-agents#232] [General]
Hi all,

As you know, I spent my last cycles adding documentation and considering good examples for the project.
It has been a lot of pain, and I'm not even sure I did a good job. Basically, the main challenge I faced is adding architectural documentation and/or examples on top of written code. While it has been pretty natural to write code and a few tests at the faster pace possible in the first interaction, I think we are already at the point we should consider documentation and examples as a required part of each new feature. 

So, starting from iteration 3 (https://github.com/orgs/wise-agents/projects/3/views/6) I think we could define a rule for each new feature implemented:

* Each new feature MUST have good docstring. So the PR reviewer should not approve a PR w/o good docstring
* Each "big-enough" feature should have proper documentation in docs. Here the question is: what does it mean "big-enough"? Size S (just XS don't need documentation), Size M? Note the documentation for those new features could be just an improvement of current documentation. The goal is to have good architectural docs and detailed explanations of "how-it-wors". Of course, any change to a class should always be reflected maintaining the docs. For this reason I think all classes which has relevant docs should link the doc in docstring.
* Regarding the previous point, we have two options: docs are part of the PR (better from many perspectives, maybe annoying if that PR is needed by anyone else to go on with his/her work), with the PR comes another issue for creating doc assigned to the same person and in the same iteration (for the cases when we need to don't block the PR)
* The same "big-enough" consideration will be relevant (maybe with different sizes). Adding non-trivial examples that use the new feature would be great for validating the feature. We have tests, but unit tests are sometimes trivial or make many assumptions. At this stage of the project, making your hands dirty and trying to use the new feature in an example could be very beneficial in general for the project

Any Thoughts?
S.

**Comment by fjuma** [comment]
> Hi all,
> 
> As you know, I spent my last cycles adding documentation and considering good examples for the project. It has been a lot of pain, and I'm not even sure I did a good job. Basically, the main challenge I faced is adding architectural documentation and/or examples on top of written code. While it has been pretty natural to write code and a few tests at the faster pace possible in the first interaction, I think we are already at the point we should consider documentation and examples as a required part of each new feature.
> 
> So, starting from iteration 3 (https://github.com/orgs/wise-agents/projects/3/views/6) I think we could define a rule for each new feature implemented:
> 
> * Each new feature MUST have good docstring. So the PR reviewer should not approve a PR w/o good docstring

+1

> * Each "big-enough" feature should have proper documentation in docs. Here the question is: what does it mean "big-enough"? Size S (just XS don't need documentation), Size M? Note the documentation for those new features could be just an improvement of current documentation. The goal is to have good architectural docs and detailed explanations of "how-it-wors". Of course, any change to a class should always be reflected maintaining the docs. For this reason I think all classes which has relevant docs should link the doc in docstring.

I think anything that adds new functionality that's user facing should have proper documentation (even if it's size is only XS).
I like the idea of linking to docs from docstrings.

> * Regarding the previous point, we have two options: docs are part of the PR (better from many perspectives, maybe annoying if that PR is needed by anyone else to go on with his/her work), with the PR comes another issue for creating doc assigned to the same person and in the same iteration (for the cases when we need to don't block the PR)

Both options seem reasonable to me. If someone wants to get more feedback on their implementation prior to working on the documentation, they could also create a draft PR to start discussions on it and once the implementation is agreed upon, they could then add the docs (or create the issue to add the docs in the same iteration if we don't want to block things) and convert from a draft PR.

> * The same "big-enough" consideration will be relevant (maybe with different sizes). Adding non-trivial examples that use the new feature would be great for validating the feature. We have tests, but unit tests are sometimes trivial or make many assumptions. At this stage of the project, making your hands dirty and trying to use the new feature in an example could be very beneficial in general for the project

Right, examples would be great to have and it's probably simplest to include the example in the same PR that adds the new feature while it's still fresh in our minds. Similar to the docs case, if we don't want to block the PR, we could also create an issue for adding the example in the same iteration.

> 
> Any Thoughts? S.

**Reply by fl4via** [reply]
I also like the idea of making the documentation mandatory as part of new features. This will allow the documentation to evolve along with the project without the need to catch up in the future, which is always a bad thing and a burden. I think that allowing it to be in a separate issue is a good idea for features that are urgent and need to be merged right away.

**Reply by maeste** [reply]
Cool, if we agree on this, We just need to take care to update CONTRIBUTING.md to reflect this new PR process. I've created an issue for this (#235 ). It's unassigned for now, but if we want to start this from the next iteration we have just tomorrow to do that. So if anyone has a spare cycle...

**Reply by fjuma** [reply]
Submitted https://github.com/wise-agents/wise-agents/pull/236
